src/table.js

```diff
@@ -684,10 +758,115 @@ export function __table(source, operations) {
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      // for json and sqlite, we already have some types, but for csv and tsv, all
+      // columns are strings here.
+      const type = typeof d[col];
+      const value = type === "string" ? d[col].trim() : d[col];
+      if (type !== "string") {
+        if (Array.isArray(value)) typeCounts[col].array++;
+        else if (value instanceof Date) typeCounts[col].date++;
+        else if (value instanceof ArrayBuffer) typeCounts[col].buffer++;
+        else if (type === "number") {
+          if (Number.isInteger(+value)) typeCounts[col].integer++;
+          else typeCounts[col].number++;
+        }
+        // bigint, boolean, or object
+        else if (type in typeCounts[col]) typeCounts[col][type]++;
+        else if (value !== null && value !== undefined) typeCounts[col].other++;
+      } else {
+        if (value === "true" || value === "false") typeCounts[col].boolean++;
+        else if (value && !isNaN(value)) {
+          if (Number.isInteger(+value)) typeCounts[col].integer++;
+          else typeCounts[col].number++;
+        } else if (/^\d+n$/.test(value)) typeCounts[col].bigint++;
+        else if (
+          value &&
+          value.match(
+            /^(([-+]\d{2})?\d{4}(-\d{1,2}(-\d{1,2})?)|(\d{1,2})\/(\d{1,2})\/(\d{2,4}))?([T ]\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/
+          )
+        )
+          typeCounts[col].date++;
+        // the long regex accepts dates in the form of ISOString and
+        // LocaleDateString, with or without times
+        else if (value) typeCounts[col].string++;
+      }
+    }
+  }
+  for (const col in typeCounts) {
+    let type = greatest(
+      Object.keys(typeCounts[col]),
+      (d) => typeCounts[col][d]
+    );
+    // If over 90% of the sampled data counted as this type, use it. Otherwise,
+    // use "other."
+    type =
+      typeCounts[col][type] / Math.min(source.length, sampleSize) >= 0.9
+        ? type
+        : "other";
```

Comment:
This is requiring 90% of the sampled values to match the most common type, but we’re including null/undefined values in the denominator. I wonder if that’s being overly strict. For example, if we sample 100 values, 50 of them are dates and 50 of them are null, it feels reasonable to say that the inferred type is date?

Reply:
Yes! Updated to use only non-nullish and non-empty string values in the denominator.